crates/sparrow-materialize/src/materialize.rs

```diff
@@ -0,0 +1,235 @@
+    /// Channel for sending the stop signal
+    stop_signal_sender: tokio::sync::watch::Sender<bool>,
+    /// Channel for receiving progress information
+    progress_receiver: tokio::sync::watch::Receiver<MaterializationProgress>,
+    /// Handle to the materialization process
+    handle: Option<JoinHandle<error_stack::Result<(), Error>>>,
+    /// Status of the materialization process
+    status: MaterializationProgress,
+}
+
+impl MaterializationControl {
+    pub fn start(
+        materialization: Materialization,
+        s3_helper: S3Helper,
+        bounded_lateness_ns: Option<i64>,
+    ) -> Self {
+        let (stop_tx, _stop_rx) = tokio::sync::watch::channel(false);
+        let (progress_tx, progress_rx) =
+            tokio::sync::watch::channel(MaterializationProgress::Uninitialized);
+
+        let handle = tokio::spawn(async move {
+            let plan = materialization.plan.clone();
+            let destination = materialization.destination.clone();
+            let tables = materialization.tables.clone();
+
+            // TODO: pass stop signal receiver
+            let mut progress_stream = sparrow_runtime::execute::materialize(
+                plan,
+                destination,
+                tables,
+                s3_helper,
+                bounded_lateness_ns,
+            )
+            .await
+            .change_context(Error::CreateMaterialization)
+            .expect("TODO") // TODO: Report error
+            .boxed();
+
+            let mut last_progress = ProgressInformation::default();
+            while let Some(message) = progress_stream
+                .try_next()
+                .await
+                .change_context(Error::ReadingProgress)?
+            {
+                if let Some(progress) = message.progress {
+                    last_progress = progress.clone();
+                    progress_tx
+                        .send(MaterializationProgress::Running(progress))
+                        .expect("TODO");
+                }
+            }
+
+            progress_tx
+                .send(MaterializationProgress::Stopped(last_progress))
+                .expect("TODO");
+            tracing::error!("materialization exited unexpectedly");
+            Ok(())
+        });
+
+        let status = MaterializationProgress::Uninitialized;
+        Self {
+            id: materialization.id.clone(),
+            stop_signal_sender: stop_tx,
+            progress_receiver: progress_rx,
+            handle: Some(handle),
+            status,
+        }
+    }
+
+    pub async fn stop(&mut self) -> error_stack::Result<(), Error> {
+        self.stop_signal_sender
+            .send(true)
+            .into_report()
+            .change_context(Error::Internal)?;
+
+        let handle = self.handle.take().expect("handle");
+        match &self.status {
+            MaterializationProgress::Uninitialized => {
+                // TODO: It's possible that the materialization process has started
```

Comment:
.

Reply:
good catch -- I guess always returning progress at the end would be a bit more correct? but probably OK either way for now.